File: online-path-learner/environment/judge.py
#!/usr/bin/env python3
"""
Judge for the Online Shortest Path problem.

Usage: python3 judge.py <seed> <solver_command...>

Generates a 30x30 grid instance from the given seed, runs the solver
interactively via stdin/stdout, and prints the final score to stdout.
Errors go to stderr.
"""

import sys
import subprocess
import random
import heapq


def dijkstra_dist(h, v, si, sj, ti, tj):
    """Compute shortest path distance from (si,sj) to (ti,tj)."""
    COLS = 30
    N = 900
    src = si * COLS + sj
    dst = ti * COLS + tj

    dist = [float('inf')] * N
    dist[src] = 0
    pq = [(0, src)]

    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        if u == dst:
            break
        ui, uj = u // COLS, u % COLS

        neighbors = []
        if uj < 29:
            neighbors.append((u + 1, h[ui][uj]))
        if uj > 0:
            neighbors.append((u - 1, h[ui][uj - 1]))
        if ui < 29:
            neighbors.append((u + COLS, v[ui][uj]))
        if ui > 0:
            neighbors.append((u - COLS, v[ui - 1][uj]))

        for nxt, w in neighbors:
            nd = d + w
            if nd < dist[nxt]:
                dist[nxt] = nd
                heapq.heappush(pq, (nd, nxt))

    return dist[dst]


def generate_instance(seed):
    """Generate a grid instance following the AHC003 specification."""
    rng = random.Random(seed)

    D = rng.randint(100, 2000)
    M = rng.randint(1, 2)

    # Horizontal edges: h[i][j] = weight of edge between (i,j) and (i,j+1)
    h = [[0] * 29 for _ in range(30)]
    for i in range(30):
        H_vals = [rng.randint(1000 + D, 9000 - D) for _ in range(M)]
        if M == 1:
            for j in range(29):
                delta = rng.randint(-D, D)
                h[i][j] = H_vals[0] + delta
        else:
            x = rng.randint(1, 28)
            for j in range(29):
                delta = rng.randint(-D, D)
                h[i][j] = H_vals[0 if j < x else 1] + delta

    # Vertical edges: v[i][j] = weight of edge between (i,j) and (i+1,j)
    v = [[0] * 30 for _ in range(29)]
    for j_col in range(30):
        V_vals = [rng.randint(1000 + D, 9000 - D) for _ in range(M)]
        if M == 1:
            for i in range(29):
                gamma = rng.randint(-D, D)
                v[i][j_col] = V_vals[0] + gamma
        else:
            y = rng.randint(1, 28)
            for i in range(29):
                gamma = rng.randint(-D, D)
                v[i][j_col] = V_vals[0 if i < y else 1] + gamma

    # Generate 1000 queries with precomputed shortest path lengths
    queries = []
    for _ in range(1000):
        while True:
            si = rng.randint(0, 29)
            sj = rng.randint(0, 29)
            ti = rng.randint(0, 29)
            tj = rng.randint(0, 29)
            if abs(si - ti) + abs(sj - tj) >= 10:
                break

        a = dijkstra_dist(h, v, si, sj, ti, tj)
        e = rng.uniform(0.9, 1.1)
        queries.append((si, sj, ti, tj, int(a), e))

    return h, v, queries, D, M


def compute_path_length(h, v, path_str, si, sj, ti, tj):
    """Validate path and compute its total edge weight.
    Returns (length, True) on success, (-1, False) on invalid path."""
    ci, cj = si, sj
    total = 0

    for ch in path_str:
        if ch == 'U':
            if ci <= 0:
                return -1, False
            ci -= 1
            total += v[ci][cj]
        elif ch == 'D':
            if ci >= 29:
                return -1, False
            total += v[ci][cj]
            ci += 1
        elif ch == 'L':
            if cj <= 0:
                return -1, False
            cj -= 1
            total += h[ci][cj]
        elif ch == 'R':
            if cj >= 29:
                return -1, False
            total += h[ci][cj]
            cj += 1
        else:
            return -1, False

    if (ci, cj) != (ti, tj):
        return -1, False

    return total, True


def main():
    if len(sys.argv) < 3:
        print("Usage: python3 judge.py <seed> <solver_command...>", file=sys.stderr)
        sys.exit(1)

    seed = int(sys.argv[1])
    solver_cmd = sys.argv[2:]

    h, v, queries, D, M = generate_instance(seed)

    proc = subprocess.Popen(
        solver_cmd,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        bufsize=1
    )

    score = 0.0

    try:
        for k, (si, sj, ti, tj, a, e) in enumerate(queries):
            # Send query to solver
            proc.stdin.write(f'{si} {sj} {ti} {tj}\n')
            proc.stdin.flush()

            # Read path from solver
            path_line = proc.stdout.readline()
            if not path_line:
                print(f"ERROR: Solver produced no output at query {k}", file=sys.stderr)
                proc.kill()
                print(0)
                return

            path_str = path_line.strip()
            b, valid = compute_path_length(h, v, path_str, si, sj, ti, tj)

            if not valid:
                print(f"ERROR: Invalid path at query {k}: "
                      f"from ({si},{sj}) to ({ti},{tj}), path='{path_str[:50]}...'",
                      file=sys.stderr)
                proc.kill()
                print(0)
                return

            # Compute noisy feedback
            feedback = round(b * e)

            # Send feedback to solver
            proc.stdin.write(f'{feedback}\n')
            proc.stdin.flush()

            # Accumulate score with exponential discount
            score = score * 0.998 + a / b

        proc.stdin.close()
        try:
            proc.wait(timeout=10)
        except subprocess.TimeoutExpired:
            proc.kill()

    except BrokenPipeError:
        print("ERROR: Solver process terminated unexpectedly", file=sys.stderr)
        try:
            proc.kill()
        except Exception:
            pass
        print(0)
        return
    except Exception as ex:
        print(f"ERROR: {ex}", file=sys.stderr)
        try:
            proc.kill()
        except Exception:
            pass
        print(0)
        return

    final_score = round(2312311 * score)
    print(final_score)


if __name__ == '__main__':
    main()

File: online-path-learner/environment/judge.c
/* Online Shortest Path Judge — compiled and stripped at build time */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdint.h>

#define NR 30
#define NC 30
#define NV (NR * NC)
#define NQ 1000
#define INF 0x7FFFFFFF

/* SplitMix64 PRNG */
typedef struct { uint64_t s; } Rng;

static uint64_t rng_next(Rng *r) {
    uint64_t z = (r->s += UINT64_C(0x9e3779b97f4a7c15));
    z = (z ^ (z >> 30)) * UINT64_C(0xbf58476d1ce4e5b9);
    z = (z ^ (z >> 27)) * UINT64_C(0x94d049bb133111eb);
    return z ^ (z >> 31);
}

static int rng_int(Rng *r, int lo, int hi) {
    return lo + (int)(rng_next(r) % (uint64_t)(hi - lo + 1));
}

static double rng_dbl(Rng *r) {
    return (rng_next(r) >> 11) * (1.0 / (double)(UINT64_C(1) << 53));
}

/* Graph edge weights */
static int hw[NR][NC - 1];  /* horizontal */
static int vw[NR - 1][NC];  /* vertical */

/* Min-heap for Dijkstra */
typedef struct { int d, n; } HNode;
static HNode heap[NV * 4];
static int hsz;

static void hpush(int d, int n) {
    int i = hsz++;
    heap[i].d = d; heap[i].n = n;
    while (i > 0) {
        int p = (i - 1) / 2;
        if (heap[p].d <= heap[i].d) break;
        HNode t = heap[p]; heap[p] = heap[i]; heap[i] = t;
        i = p;
    }
}

static HNode hpop(void) {
    HNode top = heap[0];
    heap[0] = heap[--hsz];
    int i = 0;
    for (;;) {
        int l = 2 * i + 1, r = 2 * i + 2, s = i;
        if (l < hsz && heap[l].d < heap[s].d) s = l;
        if (r < hsz && heap[r].d < heap[s].d) s = r;
        if (s == i) break;
        HNode t = heap[i]; heap[i] = heap[s]; heap[s] = t;
        i = s;
    }
    return top;
}

static int dist[NV];

static int dijkstra(int si, int sj, int ti, int tj) {
    int src = si * NC + sj, dst = ti * NC + tj;
    for (int i = 0; i < NV; i++) dist[i] = INF;
    dist[src] = 0; hsz = 0; hpush(0, src);
    while (hsz > 0) {
        HNode e = hpop();
        if (e.d > dist[e.n]) continue;
        if (e.n == dst) break;
        int ui = e.n / NC, uj = e.n % NC;
        if (uj < NC - 1) {
            int w = hw[ui][uj], nd = e.d + w;
            if (nd < dist[e.n + 1]) { dist[e.n + 1] = nd; hpush(nd, e.n + 1); }
        }
        if (uj > 0) {
            int w = hw[ui][uj - 1], nd = e.d + w;
            if (nd < dist[e.n - 1]) { dist[e.n - 1] = nd; hpush(nd, e.n - 1); }
        }
        if (ui < NR - 1) {
            int w = vw[ui][uj], nd = e.d + w;
            if (nd < dist[e.n + NC]) { dist[e.n + NC] = nd; hpush(nd, e.n + NC); }
        }
        if (ui > 0) {
            int w = vw[ui - 1][uj], nd = e.d + w;
            if (nd < dist[e.n - NC]) { dist[e.n - NC] = nd; hpush(nd, e.n - NC); }
        }
    }
    return dist[dst];
}

/* Query data */
static struct { int si, sj, ti, tj, opt; double noise; } queries[NQ];

static void generate(uint64_t seed) {
    Rng rng = { seed };
    int D = rng_int(&rng, 100, 2000);
    int M = rng_int(&rng, 1, 2);

    for (int i = 0; i < NR; i++) {
        int base[2];
        for (int p = 0; p < M; p++) base[p] = rng_int(&rng, 1000 + D, 9000 - D);
        if (M == 1) {
            for (int j = 0; j < NC - 1; j++)
                hw[i][j] = base[0] + rng_int(&rng, -D, D);
        } else {
            int x = rng_int(&rng, 1, 28);
            for (int j = 0; j < NC - 1; j++)
                hw[i][j] = base[j < x ? 0 : 1] + rng_int(&rng, -D, D);
        }
    }

    for (int j = 0; j < NC; j++) {
        int base[2];
        for (int p = 0; p < M; p++) base[p] = rng_int(&rng, 1000 + D, 9000 - D);
        if (M == 1) {
            for (int i = 0; i < NR - 1; i++)
                vw[i][j] = base[0] + rng_int(&rng, -D, D);
        } else {
            int y = rng_int(&rng, 1, 28);
            for (int i = 0; i < NR - 1; i++)
                vw[i][j] = base[i < y ? 0 : 1] + rng_int(&rng, -D, D);
        }
    }

    for (int q = 0; q < NQ; q++) {
        int si, sj, ti, tj;
        do {
            si = rng_int(&rng, 0, 29); sj = rng_int(&rng, 0, 29);
            ti = rng_int(&rng, 0, 29); tj = rng_int(&rng, 0, 29);
        } while (abs(si - ti) + abs(sj - tj) < 10);
        queries[q].si = si; queries[q].sj = sj;
        queries[q].ti = ti; queries[q].tj = tj;
        queries[q].opt = dijkstra(si, sj, ti, tj);
        queries[q].noise = 0.9 + 0.2 * rng_dbl(&rng);
    }
}

static int path_cost(const char *path, int si, int sj, int ti, int tj) {
    int ci = si, cj = sj, total = 0;
    for (int k = 0; path[k] && path[k] != '\n' && path[k] != '\r'; k++) {
        switch (path[k]) {
            case 'U': if (ci <= 0) return -1; ci--; total += vw[ci][cj]; break;
            case 'D': if (ci >= NR - 1) return -1; total += vw[ci][cj]; ci++; break;
            case 'L': if (cj <= 0) return -1; cj--; total += hw[ci][cj]; break;
            case 'R': if (cj >= NC - 1) return -1; total += hw[ci][cj]; cj++; break;
            default: return -1;
        }
    }
    return (ci == ti && cj == tj) ? total : -1;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <seed> <cmd...>\n", argv[0]);
        return 1;
    }
    signal(SIGPIPE, SIG_IGN);
    generate(strtoull(argv[1], NULL, 10));

    int p2c[2], c2p[2];
    if (pipe(p2c) < 0 || pipe(c2p) < 0) { perror("pipe"); return 1; }
    pid_t pid = fork();
    if (pid < 0) { perror("fork"); return 1; }

    if (pid == 0) {
        close(p2c[1]); close(c2p[0]);
        dup2(p2c[0], STDIN_FILENO);
        dup2(c2p[1], STDOUT_FILENO);
        close(p2c[0]); close(c2p[1]);
        execvp(argv[2], argv + 2);
        perror("exec"); _exit(1);
    }

    close(p2c[0]); close(c2p[1]);
    FILE *wr = fdopen(p2c[1], "w");
    FILE *rd = fdopen(c2p[0], "r");
    if (!wr || !rd) { fprintf(stderr, "fdopen failed\n"); return 1; }

    char buf[4096];
    double score = 0.0;

    for (int k = 0; k < NQ; k++) {
        fprintf(wr, "%d %d %d %d\n",
                queries[k].si, queries[k].sj,
                queries[k].ti, queries[k].tj);
        fflush(wr);

        if (!fgets(buf, sizeof(buf), rd)) {
            fprintf(stderr, "No output at query %d\n", k);
            printf("0\n");
            kill(pid, SIGKILL); waitpid(pid, NULL, 0);
            return 0;
        }
        int len = strlen(buf);
        while (len > 0 && (buf[len - 1] == '\n' || buf[len - 1] == '\r'))
            buf[--len] = '\0';

        int b = path_cost(buf, queries[k].si, queries[k].sj,
                          queries[k].ti, queries[k].tj);
        if (b < 0) {
            fprintf(stderr, "Invalid path at query %d\n", k);
            printf("0\n");
            kill(pid, SIGKILL); waitpid(pid, NULL, 0);
            return 0;
        }

        int fb = (int)(b * queries[k].noise + 0.5);
        fprintf(wr, "%d\n", fb);
        fflush(wr);

        score = score * 0.998 + (double)queries[k].opt / (double)b;
    }

    fclose(wr);
    waitpid(pid, NULL, 0);
    fclose(rd);

    printf("%lld\n", (long long)(2312311.0 * score + 0.5));
    return 0;
}
